pennylane_lightning/src/simulator/KernelMap.hpp

```diff
@@ -0,0 +1,322 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Set/get Default kernels for statevector
+ */
+#include "Error.hpp"
+#include "GateOperation.hpp"
+#include "IntegerInterval.hpp"
+#include "KernelType.hpp"
+#include "Threading.hpp"
+#include "Util.hpp"
+
+#include <deque>
+#include <functional>
+#include <unordered_map>
+#include <utility>
+
+namespace Pennylane::KernelMap {
+///@cond DEV
+namespace Internal {
+
+int assignDefaultKernelsForGateOp();
+int assignDefaultKernelsForGeneratorOp();
+int assignDefaultKernelsForMatrixOp();
+
+template <class Operation> struct AssignKernelForOp;
+
+template <> struct AssignKernelForOp<Gates::GateOperation> {
+    static inline const int dummy = assignDefaultKernelsForGateOp();
+};
+template <> struct AssignKernelForOp<Gates::GeneratorOperation> {
+    static inline const int dummy = assignDefaultKernelsForGeneratorOp();
+};
+template <> struct AssignKernelForOp<Gates::MatrixOperation> {
+    static inline const int dummy = assignDefaultKernelsForMatrixOp();
+};
+} // namespace Internal
+///@endcond
+
+///@cond DEV
+struct DispatchElement {
+    uint32_t priority;
+    Util::IntegerInterval<size_t> interval;
+    Gates::KernelType kernel;
+};
+
+inline bool lower_priority(const DispatchElement &lhs,
+                           const DispatchElement &rhs) {
+    return lhs.priority < rhs.priority;
+}
+
+inline bool higher_priority(const DispatchElement &lhs,
+                            const DispatchElement &rhs) {
+    return lhs.priority > rhs.priority;
+}
+
+/**
+ * @brief Maintain dispatch element using a vector decreasingly-ordered by
+ * priority.
+ */
+class PriorityDispatchSet {
```

Comment:
Should we use `https://en.cppreference.com/w/cpp/algorithm/make_heap` here to handle the priority ordering after insertion?

Reply:
Heap (priority queue) works when we only want the highest priority object. Here I want to pick up the object with the highest priority among the object with the given pair of (number of wires, threading option). So I think it is not straightforward to use a heap (and using the sorted array has the same time complexity). Still, I expect that the kernel map mechanism may be rewritten when we have more kernels (AVX, OpenMP, and maybe Kokkos).

Reply:
Ah, ok, I see now. Sure, that's fine.